moveit_setup_assistant/src/widgets/simulation_widget.cpp

```diff
@@ -70,161 +73,236 @@ SimulationWidget::SimulationWidget(QWidget* parent, const MoveItConfigDataPtr& c
+      continue;  // ignore invalid tags
+    if (std::string(type_tag->GetText()) == "transmission_interface/SimpleTransmission")
+      transmission_elements[element->FirstChildElement("joint")->Attribute("name")] = element;
+  }
+
+  // Loop through Link and Joint elements and add Gazebo tags if not present
+  for (TiXmlElement* element = root->FirstChildElement(); element != nullptr; element = element->NextSiblingElement())
+  {
+    const std::string tag_name(element->Value());
+    if (tag_name == "link" && element->FirstChildElement("collision"))
+    {
+      TiXmlElement* inertial = uniqueInsert(*element, "inertial");
+      uniqueInsert(*inertial, "mass", { { "value", "0.1" } });
+      uniqueInsert(*inertial, "origin", { { "xyz", "0 0 0" }, { "rpy", "0 0 0" } });
+      uniqueInsert(*inertial, "inertia",
+                   { { "ixx", "0.03" },
+                     { "iyy", "0.03" },
+                     { "izz", "0.03" },
+                     { "ixy", "0.0" },
+                     { "ixz", "0.0" },
+                     { "iyz", "0.0" } });
+    }
+    else if (tag_name == "joint")
+    {
+      const char* joint_type = element->Attribute("type");
+      const char* joint_name = element->Attribute("name");
+      if (!joint_type || !joint_name || strcmp(joint_type, "fixed") == 0)
+        continue;  // skip invalid or fixed joints
+
+      // find existing or create new transmission element for this joint
+      TiXmlElement* transmission;
+      auto it = transmission_elements.find(joint_name);
+      if (it != transmission_elements.end())
+        transmission = it->second;
+      else
+      {
+        transmission = root->InsertEndChild(TiXmlElement("transmission"))->ToElement();
+        transmission->SetAttribute("name", std::string("trans_") + joint_name);
+      }
+
+      uniqueInsert(*transmission, "type", {}, "transmission_interface/SimpleTransmission");
+
+      std::string hw_interface = config_data_->getJointHardwareInterface(joint_name);
+      auto* joint = uniqueInsert(*transmission, "joint", { { "name", joint_name } });
+      uniqueInsert(*joint, "hardwareInterface", {}, hw_interface.c_str());
+
+      auto actuator_name = joint_name + std::string("_motor");
+      auto* actuator = uniqueInsert(*transmission, "actuator", { { "name", actuator_name.c_str() } });
+      uniqueInsert(*actuator, "hardwareInterface", {}, hw_interface.c_str());
+      uniqueInsert(*actuator, "mechanicalReduction", {}, "1");
+    }
+  }
+
+  // Add gazebo_ros_control plugin which reads the transmission tags
+  TiXmlElement* gazebo = uniqueInsert(*root, "gazebo");
+  TiXmlElement* plugin = uniqueInsert(*gazebo, "plugin", { { "name", "gazebo_ros_control", true } });
+  uniqueInsert(*plugin, "robotNamespace", {}, "/");
+
+  // generate new URDF
+  TiXmlPrinter new_urdf;
+  doc.Accept(&new_urdf);
+  // and return it when there are changes
+  return orig_urdf.Str() == new_urdf.Str() ? std::string() : new_urdf.Str();
```

Comment:
@rhaschke This check now works for the majority of the input URDFs. This check fails for the panda robot description since @frankaemika uses non-standard formatting for the `inertia` tag. This formatting is removed when we parse the URDF using the [tinyxml](http://www.grinninglizard.com/tinyxml/) library. I, however, think complicating our code to handle this unique use case might not be worth it. I like the new version more anyway.

**franka inertia**

```yaml
<inertial>
  <mass value="0.73" />
  <origin xyz="-0.01 0 0.03" />
  <inertia ixx="0.001" ixy="0"      ixz="0"
                        iyy="0.0025" iyz="0"
                                    izz="0.0017" />
</inertial>
```
**the inertia tag we return**

```yml
<inertia ixx="0.001" ixy="0" ixz="0" iyy="0.0025" iyz="0" izz="0.0017" />
```

Reply:
I agree. Formatting this nicely isn't worth the effort. By the way, if there are no changes to the URDF, the formatting is kept as well.